### PR TITLE
Use a new method for handling soft masks.

### LINKED ISSUE #5781: Unable to render all pdf contents
When I am trying to load this pdf http://merlin_test.s3.amazonaws.com/myPdf.pdf, it's unable to render all the contents. There is a blue background which is not visible and some of the texts are not visible as well as the color is not proper. Can someone please help

### LINKED ISSUE #5853: Pdf Render with Worng colors, Missing Images and Elements
hello,
this pdf file created with Adobe InDesign, alot of elements are missing, also wrong color are displaying:
http://tal-ram.co.il/images/products/orders/14265798291.pdf

viewer

org

### LINKED ISSUE #6573: PDF does not complete loading, stuck with spinner
PDFs from lumileds.com do not load with PDF.js. Part of the background loads, then only the spinner continues while using 100% CPU. 

An example pdf is http://www.lumileds.com/uploads/17/DS61-pdf 

This behaviour occurs with both the release version of pdf.js in firefox 41.0.2 and the latest .xpi downloaded from this github page.

### LINKED ISSUE #7267: PDF.js fails to create complete/correct preview
**Link to PDF file (or attach file here):**
[tocheckFile.pdf](https://github.com/mozilla/pdf.js/files/240714/tocheckFile.pdf)

**Configuration:**
- Web browser and its version: FF45.0.2, Chrome Version 50.0.2661.86 (64-bit) 
- Operating system and its version: Mac OSX El Capitan

**- PDF.js version:**
PDFJS.version = '1.1.215';
PDFJS.build = 'c9a7498';

**Steps to reproduce the problem:**
1. Use the lib to generate a PDF preview.

**What is the expected behavior? (add screenshot)**
PDF preview should have rendered completely. (frog head should be complete)

**What went wrong? (add screenshot)**
imagesnotrenderinginpreview

See attached screenshot in regards to the pdf file

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):

### LINKED ISSUE #7842: greyscale gradient does not display correctly
Link to PDF file (or attach file here): 
[gradient_pdfjs.pdf](https://github.com/mozilla/pdf.js/files/607650/gradient_pdfjs.pdf)

Configuration:
- Web browser and its version: Chrome 54.0.2840.98
- Operating system and its version: Mac OSX 10.11.6
- PDF.js version: 1.6.210
- Is an extension: No

Steps to reproduce the problem:
1. View the PDF with PDFJS

What is the expected behavior? (add screenshot)

image

What went wrong? (add screenshot)

image

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension): reproduced at https://mozilla.github.io/pdf.js/web/viewer.html

### LINKED ISSUE #7891: softmask backdrop ignored
Link to PDF file (or attach file here):

[gs-bugzilla689309-reduced-bc1_RGB.pdf](https://github.com/mozilla/pdf.js/files/649453/gs-bugzilla689309-reduced-bc1_RGB.pdf)
[gs-bugzilla689309-reduced-bc0_RGB.pdf](https://github.com/mozilla/pdf.js/files/649454/gs-bugzilla689309-reduced-bc0_RGB.pdf)

Configuration:
- Web browser and its version: ff 50.0.2
- Operating system and its version: W7 64bit fully patched
- PDF.js version: most recent
- Is an extension: ?

Steps to reproduce the problem:
1. open demo viewer
2. open file gs-bugzilla689309-reduced-bc1_RGB.pdf

What is the expected behavior? (add screenshot)
screenshot-adobe

What went wrong? (add screenshot)
screenshot-pdfjs

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):
n/a

Additional comment:
The two files are identical except for the /bc value in the softmask. The softmask is smaller than the transparency group it is applied to. My understanding is that the /bc color is to be used instead of the softmask as an "influence" for the area that isn't covered by the softmask. I suspect that PDF.js just ignores /bc.

The fine black line doesn't belong there (although it happens to be useful as it shows the size of the softmask). The red line does belong there and shows the size of the transparency group.

The file is a reduced and modified version of one in https://bugs.ghostscript.com/show_bug.cgi?id=688601 to investigate a different problem than the one mentioned there.

### LINKED ISSUE #8179: Some parts missing from an image, and wrong size
This PDF has an image towards the top of the first page, and it displays incorrectly in the latest pdf.js version. This seems to be the case on all browsers we've tried, including Chrome/Firefox latest versions on Mac in my case.

It was created using CS5: Adobe InDesign, saved as a "high Quality Print".

[2017 Flyer_Mailer_W vines Logo-optimized.pdf](https://github.com/mozilla/pdf.js/files/857754/2017.Flyer_Mailer_W.vines.Logo-optimized.pdf)

What is the expected behavior? (add screenshot)

screen shot 2017-03-21 at 10 19 51

What went wrong? (add screenshot)

screen shot 2017-03-21 at 10 19 42

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):

https://mozilla.github.io/pdf.js/web/viewer.html

Thank you!

Dan

### LINKED ISSUE #8403: Slow rendering/Failure to render font
Link to PDF file (or attach file here):
[RP_25884_V2_UEL Semi Finals 2016-17_concepts.pdf](https://github.com/mozilla/pdf.js/files/994423/RP_25884_V2_UEL.Semi.Finals.2016-17_concepts.pdf)

Configuration:
- Web browser and its version: Chrome 58, Firefox 53
- Operating system and its version: MacOS 10.11.6
- PDF.js version:
- Is an extension: No

Steps to reproduce the problem:
1.  View the file the file

What is the expected behavior? (add screenshot)
screen shot 2017-05-11 at 2 06 36 pm

What went wrong? (add screenshot)
Textlayer exists here but font fails to render. No warning in the console from what I can see. 

screen shot 2017-05-11 at 2 06 20 pm

Thanks!

### LINKED ISSUE #8624: PDFjs don't render drop shadows
Link to PDF file (or attach file here):
[pages4.pdf](https://github.com/mozilla/pdf.js/files/1128300/pages4.pdf)

Configuration:
- Web browser and its version: Firefox 54
- Operating system and its version: Win 10
- PDF.js version: 1.8.522
- Is an extension: no

What is the expected behavior? (add screenshot)
chrome1-1 1
chrome1-2 1

What went wrong? (add screenshot)
screenshot from 2017-07-06 12-10-30
screenshot from 2017-07-06 12-10-44

### LINKED ISSUE #9017: PDF Renders Badly
Attach (recommended) or Link to PDF file here: https://www.recom-power.com/pdf/Powerline-AC-DC/RAC05-K.pdf

Configuration:
- Web browser and its version: Firefox 56.0 (64-bit)
- Operating system and its version: Windows 10 with OS Build 15063.632
- PDF.js version: ?
- Is an extension: Built-in

Steps to reproduce the problem:
1. Open Page

What is the expected behavior? To properly render the first page. 
image

What went wrong? Renders poorly.
image

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension): ?

### PR DESCRIPTION
**Pull request by brendandahl**
Commit 1:
The old method of handling soft masks had a number of issues where the temporary
drawing canvas and the suspended main canvas could get out of sync
(e.g. mismatched save/restores or clip state) or we could end up compositing at
the wrong time. A good example of things getting out sync is the reduced test
case in #9017.

To fix this I've changed two big things:

1) Duplicate all the needed graphics state from the temporary canvas to the
suspended main canvas. This ensure the canvases stay in sync so that when we
switch back to the main canvas the graphics state stack is the same
(e.g. transforms, clip paths).

2) Immediately composite after each drawing operation. This ensures that if
there's an active clip region that we'll still be able to composite the correct
portions of the canvas. Note: This solution could be avoided by using
getImageData and putImageData since those ignore clipping region, but this is
very very slow. Note2: I also think the old way of only compositing at the end
of the soft mask is incorrect and can lead to wrong colors if drawing over the
same region, but in practice this doesn't seem to matter much.

Fixes: #5781
Fixes: #5853
Fixes: #7267
Fixes: #7891
Fixes: #8403
Fixes: #8624
Fixes: #12798
Fixes: #13891
Fixes: #9017 (reduced test case)
Fixes: https://bugzilla.mozilla.org/show_bug.cgi?id=1703683

Commit 2:
This allows us to compose much smaller regions of soft
mask making them much faster. This should also allow
for further optimizations in the pattern code.

For example locally I see issue #6573 go from 55s
to 5s with this change.

Fixes #6573

### CODE REVIEW

File: src/core/evaluator.js [hunk 1]
```diff
@@ -315,6 +315,9 @@ class PartialEvaluator {
               }
             }
           }
```
Comment (brendandahl) [inline src/core/evaluator.js:317]: I don't think this is needed anymore. Will double check locally.

**Review by Snuffleupagus — COMMENTED** [review]
I've made a first pass through the code, and I must say that all of the added comments are most helpful here!

One small thing that I noticed is that the test-case from https://bugzilla.mozilla.org/show_bug.cgi?id=986450 still renders incorrectly even with this patch (on Windows at least).
Looking at the test-results, there also appears to be a *small* regression (when compared with Adobe Reader) in the `bug1123803` test-case on all platforms/browsers.

**Review by Snuffleupagus — APPROVED** [review]
r=me, with a final test-run to make sure that everything still looks as it should.
This looks like a *really* good improvement, w.r.t. both rendering and performance; thank you!

**Pull request merged**